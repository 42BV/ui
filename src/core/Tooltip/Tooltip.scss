.rc-tooltip {
  opacity: 1;
}
