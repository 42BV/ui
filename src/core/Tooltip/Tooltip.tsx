import React from 'react';
import Tippy from '@tippyjs/react';
import { ToolTipProps } from '../types';

/**
 * Bootstrap-like Tooltip component based on the Tippy.js library.
 */
export function Tooltip({
  children,
  placement = 'top',
  content,
  offset = 0,
  distance = 7,
  interactive,
  maxWidth = 350,
  tag = 'span',
  className,
  style
}: ToolTipProps) {
  const Tag = tag;
  const outline = !!style?.outline ? style.outline : 0;

  return (
    <Tippy
      className="border-0"
      content={content}
      placement={placement}
      offset={[offset, distance]}
      interactive={interactive}
      maxWidth={maxWidth}
    >
      <Tag
        className={className}
        style={{ outline, ...style }}
        tabIndex={0}
        role="button"
      >
        {children}
      </Tag>
    </Tippy>
  );
}
