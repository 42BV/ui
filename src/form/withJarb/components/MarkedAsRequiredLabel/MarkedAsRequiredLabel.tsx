import React from 'react';
import { isRequired } from '@42.nl/jarb-final-form';
import { getConfig } from '../../../../config/config';
import { t } from '../../../../utilities/translation/translation';

type Props = {
  label?: React.ReactNode;
  validator: string;
};

export function MarkedAsRequiredLabel({ label, validator }: Props) {
  if (!label) {
    return null;
  }

  const { showRequiredMarkInLabel } = getConfig();

  if (
    typeof label === 'string' &&
    showRequiredMarkInLabel &&
    isRequired(validator)
  ) {
    return (
      <>
        {label}
        {t({
          key: 'withJarb.REQUIRED_MARK',
          fallback: ' *'
        })}
      </>
    );
  }

  return <>{label}</>;
}
