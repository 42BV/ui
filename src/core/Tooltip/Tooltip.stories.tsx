import React from 'react';
import { storiesOf } from '@storybook/react';

import Tooltip from './Tooltip';
import Tag from '../Tag/Tag';
import { Button, Col, Row } from 'reactstrap';

storiesOf('core|Tooltip', module)
  .addParameters({ component: Tooltip })
  .add('default', () => (
    <div className="d-flex flex-column">
      <Row className="my-3">
        <Col className="d-flex justify-content-around align-items-center">
          <Tooltip content="This is a tooltip">
            <Tag color="danger" text="Hover me!" />
          </Tooltip>

          <Tooltip content="Tooltip-text">
            <h5>Can be wrapped around any component</h5>
          </Tooltip>

          <Tooltip content="This is a tooltip">Plain text</Tooltip>

          <Tooltip content="Buttons work too">
            <Button>Hover this button!</Button>
          </Tooltip>
        </Col>
      </Row>
    </div>
  ))

  .add('alignment', () => (
    <>
      <h6>Alignment</h6>
      <Row className="mt-4">
        <Col className="d-flex justify-content-around">
          <Tooltip content={'Placement bottom'} placement="bottom">
            <Tag color="danger" text="Hover me!" />
          </Tooltip>

          <Tooltip content={'Placement left'} placement="left">
            <Tag color="warning" text="Hover me!" />
          </Tooltip>
          <Tooltip content={'Placement right'} placement="right">
            <Tag color="primary" text="Hover me!" />
          </Tooltip>

          <Tooltip content={'Placement top'} placement="top">
            <Tag color="success" text="Hover me!" />
          </Tooltip>
        </Col>
      </Row>
      <hr />

      <h6>Alignment-modifier</h6>
      <Row className="mt-3">
        <Col className="d-flex justify-content-around">
          <Tooltip content={'right-start'} placement="right-start">
            <Button style={{ height: 75 }}> Hover me! </Button>
          </Tooltip>

          <Tooltip content={'right'} placement="right">
            <Button style={{ height: 75 }}> Hover me! </Button>
          </Tooltip>

          <Tooltip content={'right-end'} placement="right-end">
            <Button style={{ height: 75 }}> Hover me! </Button>
          </Tooltip>
        </Col>
      </Row>
    </>
  ))

  .add('width & distance', () => (
    <>
      <h6>Max width</h6>
      <Row className="mt-4">
        <Col className="d-flex justify-content-around">
          <Tooltip
            content={<p className="text-center">You can set me to be narrow</p>}
            maxWidth={80}
          >
            <Tag color="warning" text={'Hover me!'} />
          </Tooltip>

          <Tooltip
            content={
              <p className="text-center">Or you can set me to be wide...</p>
            }
            maxWidth={500}
          >
            <Tag color="warning" text={'Hover me!'} />
          </Tooltip>
        </Col>
      </Row>

      <hr />

      <h6>Distance</h6>
      <Row className="mt-4">
        <Col className="d-flex justify-content-around">
          <Tooltip content="far away" distance={15}>
            <Tag color="success" text="Hover me!" />
          </Tooltip>

          <Tooltip content="default distance">
            <Tag color="success" text="Hover me!" />
          </Tooltip>

          <Tooltip content="very close" distance={3}>
            <Tag color="success" text="Hover me!" />
          </Tooltip>
        </Col>
      </Row>
    </>
  ))

  .add('components as content', () => (
    <>
      <h6>Components</h6>
      <Row className="mt-4">
        <Col className="d-flex justify-content-around">
          <Tooltip
            interactive
            content={
              <div>
                {' '}
                I can render <b style={{ color: 'orange' }}> HTML content</b>
              </div>
            }
          >
            <p color="orange">HTML</p>
          </Tooltip>

          <Tooltip
            interactive
            content={<Button color="danger"> React components too!</Button>}
          >
            <p>React</p>
          </Tooltip>

          <Tooltip
            interactive
            placement="left"
            content={
              <Tooltip
                interactive
                placement="left"
                content={
                  <Tooltip
                    interactive
                    placement="left"
                    content={
                      <Tooltip
                        interactive
                        placement="left"
                        content={
                          <Tooltip
                            interactive
                            placement="left"
                            content={
                              <Tooltip interactive content="Pretty cool right?">
                                <p>5</p>
                              </Tooltip>
                            }
                          >
                            <p>4</p>
                          </Tooltip>
                        }
                      >
                        <p>3</p>
                      </Tooltip>
                    }
                  >
                    <p>2</p>
                  </Tooltip>
                }
              >
                <p>1</p>
              </Tooltip>
            }
          >
            <p>Tooltip-ception</p>
          </Tooltip>
        </Col>
      </Row>

      <hr />
      <h6>Interactive</h6>

      <Row className="mt-4">
        <Col className="d-flex justify-content-around">
          <Tooltip content="You cannot click my contents">
            <Tag color="danger" text="I am not interactive" />
          </Tooltip>
          <Tooltip interactive content="You can click my content!">
            <Tag color="success" text="I am interactive!" />
          </Tooltip>
        </Col>
      </Row>
    </>
  ))

  .add('custom wrapper', () => (
    <>
      <h6>Custom wrapper</h6>
      <Row className="mt-4">
        <Col className="d-flex justify-content-around">
          <Tooltip content="My children are in a <span>" tag="span">
            By default, my children get wrapped in a span
          </Tooltip>

          <Tooltip content="My children are in a <div>" tag="div">
            You can change that with the tag property
          </Tooltip>
        </Col>
      </Row>
    </>
  ));
