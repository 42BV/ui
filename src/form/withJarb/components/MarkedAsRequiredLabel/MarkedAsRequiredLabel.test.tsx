import React from 'react';
import * as jarbFinalForm from '@42.nl/jarb-final-form';
import { MarkedAsRequiredLabel } from './MarkedAsRequiredLabel';
import { configure } from '../../../..';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Tooltip from '../../../../core/Tooltip/Tooltip';
import { setConstraints } from '@42.nl/jarb-final-form';

describe('Component: MarkedAsRequiredLabel', () => {
  function setup({
    hasLabel = true,
    customLabel = false
  }: {
    hasLabel?: boolean;
    customLabel?: boolean;
  }) {
    const markedAsRequiredLabel = shallow(
      <MarkedAsRequiredLabel
        validator="Person.firstName"
        label={
          hasLabel ? (
            customLabel ? (
              <Tooltip content="This field is required">First Name</Tooltip>
            ) : (
              'First Name'
            )
          ) : (
            undefined
          )
        }
      />
    );
    return { markedAsRequiredLabel };
  }

  it('should not render anything when label is undefined', () => {
    const { markedAsRequiredLabel } = setup({ hasLabel: false });
    expect(toJson(markedAsRequiredLabel)).toMatchSnapshot(
      'Component: MarkedAsRequiredLabel => should return undefined when label is undefined'
    );
  });

  it('should render label without required mark when showRequiredMarkInLabel is disabled in config', () => {
    configure({ showRequiredMarkInLabel: false });
    const { markedAsRequiredLabel } = setup({});
    expect(toJson(markedAsRequiredLabel)).toMatchSnapshot(
      'Component: MarkedAsRequiredLabel => should render label without required mark when showRequiredMarkInLabel is disabled in config'
    );
    configure({ showRequiredMarkInLabel: true });
  });

  it('should render label without required mark when constraints are not defined', () => {
    jest.spyOn(jarbFinalForm, 'getConstraints').mockReturnValue(undefined);
    const { markedAsRequiredLabel } = setup({});
    expect(toJson(markedAsRequiredLabel)).toMatchSnapshot(
      'Component: MarkedAsRequiredLabel => should render label without required mark when constraints are not defined'
    );
  });

  it('should render label without required mark when constraints does not include field validator', () => {
    jest.spyOn(jarbFinalForm, 'getConstraints').mockReturnValue({ User: {} });
    const { markedAsRequiredLabel } = setup({});
    expect(toJson(markedAsRequiredLabel)).toMatchSnapshot(
      'Component: MarkedAsRequiredLabel => should render label without required mark when constraints does not include field validator'
    );
  });

  it('should render label without required mark when constraints include field validator but field is not required', () => {
    jest.spyOn(jarbFinalForm, 'getConstraints').mockReturnValue({
      Person: {
        firstName: {
          javaType: 'java.lang.String',
          types: ['text'],
          required: false,
          minimumLength: null,
          maximumLength: 50,
          fractionLength: null,
          radix: null,
          pattern: null,
          min: null,
          max: null,
          name: 'name'
        }
      }
    });
    const { markedAsRequiredLabel } = setup({});
    expect(toJson(markedAsRequiredLabel)).toMatchSnapshot(
      'Component: MarkedAsRequiredLabel => should render label without required mark when constraints include field validator but field is not required'
    );
  });

  it('should render label without required mark when label is a custom component', () => {
    jest.spyOn(jarbFinalForm, 'getConstraints').mockReturnValue({
      Person: {
        firstName: {
          javaType: 'java.lang.String',
          types: ['text'],
          required: false,
          minimumLength: null,
          maximumLength: 50,
          fractionLength: null,
          radix: null,
          pattern: null,
          min: null,
          max: null,
          name: 'name'
        }
      }
    });
    const { markedAsRequiredLabel } = setup({ customLabel: true });
    expect(toJson(markedAsRequiredLabel)).toMatchSnapshot(
      'Component: MarkedAsRequiredLabel => should render label without required mark when label is a custom component'
    );
  });

  it('should render label with required mark when constraints include field validator', () => {
    setConstraints({
      Person: {
        firstName: {
          javaType: 'java.lang.String',
          types: ['text'],
          required: true,
          minimumLength: null,
          maximumLength: 50,
          fractionLength: null,
          radix: null,
          pattern: null,
          min: null,
          max: null,
          name: 'name'
        }
      }
    });
    const { markedAsRequiredLabel } = setup({});
    expect(toJson(markedAsRequiredLabel)).toMatchSnapshot(
      'Component: MarkedAsRequiredLabel => should render label with required mark when constraints include field validator'
    );
  });
});
