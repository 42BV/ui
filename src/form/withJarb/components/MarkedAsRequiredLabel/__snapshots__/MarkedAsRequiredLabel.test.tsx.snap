// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: MarkedAsRequiredLabel should not render anything when label is undefined: Component: MarkedAsRequiredLabel => should return undefined when label is undefined 1`] = `""`;

exports[`Component: MarkedAsRequiredLabel should render label with required mark when constraints include field validator: Component: MarkedAsRequiredLabel => should render label with required mark when constraints include field validator 1`] = `
<Fragment>
  First Name
   *
</Fragment>
`;

exports[`Component: MarkedAsRequiredLabel should render label without required mark when constraints are not defined: Component: MarkedAsRequiredLabel => should render label without required mark when constraints are not defined 1`] = `
<Fragment>
  First Name
</Fragment>
`;

exports[`Component: MarkedAsRequiredLabel should render label without required mark when constraints does not include field validator: Component: MarkedAsRequiredLabel => should render label without required mark when constraints does not include field validator 1`] = `
<Fragment>
  First Name
</Fragment>
`;

exports[`Component: MarkedAsRequiredLabel should render label without required mark when constraints include field validator but field is not required: Component: MarkedAsRequiredLabel => should render label without required mark when constraints include field validator but field is not required 1`] = `
<Fragment>
  First Name
</Fragment>
`;

exports[`Component: MarkedAsRequiredLabel should render label without required mark when label is a custom component: Component: MarkedAsRequiredLabel => should render label without required mark when label is a custom component 1`] = `
<Fragment>
  <Tooltip
    content="This field is required"
  >
    First Name
  </Tooltip>
</Fragment>
`;

exports[`Component: MarkedAsRequiredLabel should render label without required mark when showRequiredMarkInLabel is disabled in config: Component: MarkedAsRequiredLabel => should render label without required mark when showRequiredMarkInLabel is disabled in config 1`] = `
<Fragment>
  First Name
</Fragment>
`;
